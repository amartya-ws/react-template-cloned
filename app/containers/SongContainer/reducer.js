/*
 *
 * HomeContainer reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';

export const { Types: songContainerTypes, Creators: songContainerCreators } = createActions({
  requestGetItuneSongs: ['artistName'],
  successGetItuneSongs: ['data'],
  clearItuneSongs: {}
});
export const initialState = { artistName: null, ituneData: {} };

/* eslint-disable default-case, no-param-reassign */
export const songContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case songContainerTypes.REQUEST_GET_ITUNE_SONGS:
        draft.artistName = action.artistName;
        break;
      case songContainerTypes.SUCCESS_GET_ITUNE_SONGS:
        draft.ituneData = action.data;
        break;
      case songContainerTypes.CLEAR_ITUNE_SONGS:
        draft.artistName = null;
        draft.ituneData = {};
        break;
    }
  });

export default songContainerReducer;
