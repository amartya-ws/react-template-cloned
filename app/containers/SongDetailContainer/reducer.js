/*
 *
 * SongDetail reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';

export const initialState = {
  trackId: null,
  singleItune: {}
};

export const { Types: songDetailContainerTypes, Creators: songDetailContainerCreators } = createActions({
  requestGetSingleItune: ['trackId'],
  successGetSingleItune: ['data']
});

/* eslint-disable default-case, no-param-reassign */
export const songDetailContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case songDetailContainerTypes.REQUEST_GET_SINGLE_ITUNE:
        draft.trackId = action.trackId;
        break;
      case songDetailContainerTypes.SUCCESS_GET_SINGLE_ITUNE:
        draft.singleItune = action.data;
        break;
    }
  });

export default songDetailContainerReducer;
