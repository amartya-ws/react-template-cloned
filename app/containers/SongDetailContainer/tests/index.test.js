/**
 *
 * Tests for SongDetail container
 *
 *
 */

import React from 'react';
import { renderProvider, timeout } from '@utils/testUtils';
import { SongDetailContainerTest as SongDetailContainer, mapDispatchToProps } from '../index';
import { songDetailContainerTypes } from '../reducer';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(() => ({ trackId: 12345 }))
}));

describe('<SongDetailContainer /> container tests', () => {
  let submitSpy;

  beforeEach(() => {
    submitSpy = jest.fn();
    //jest.spyOn(Router, 'useParams').mockReturnValue({ trackId: 12345 });
  });
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<SongDetailContainer />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchGetSingleItune when data is not available', async () => {
    let trackId = 12345;
    let ituneData = {};
    let singleItune = {};

    renderProvider(
      <SongDetailContainer ituneData={ituneData} singleItune={singleItune} dispatchGetSingleItune={submitSpy} />
    );

    await timeout(500);
    expect(submitSpy).toBeCalledWith(trackId);
  });

  it('should should not call dispatchGetSingleItune when ituneData is available', async () => {
    let ituneData = {
      resultCount: 1,
      results: [
        { trackId: 12345, artistName: 'tupac', trackName: 'california love', artworkUrl100: 'https://demo.com' }
      ]
    };
    let singleItune = {};

    const { getByTestId } = renderProvider(
      <SongDetailContainer ituneData={ituneData} singleItune={singleItune} dispatchGetSingleItune={submitSpy} />
    );

    await timeout(500);
    expect(submitSpy).not.toBeCalled();
    expect(getByTestId('song-image')).toHaveAttribute('src', ituneData.results[0].artworkUrl100);
  });

  it('should should not call dispatchGetSingleItune when data is available in singleData', async () => {
    let ituneData = {};
    let singleItune = {
      resultCount: 1,
      results: [
        { trackId: 12345, artistName: 'tupac', trackName: 'california love', artworkUrl100: 'https://demo.com' }
      ]
    };

    const { getByTestId } = renderProvider(
      <SongDetailContainer ituneData={ituneData} singleItune={singleItune} dispatchGetSingleItune={submitSpy} />
    );

    await timeout(500);
    expect(submitSpy).not.toBeCalled();
    expect(getByTestId('song-image')).toHaveAttribute('src', singleItune.results[0].artworkUrl100);
  });

  it('should validate mapDispatchToProps actions', async () => {
    const trackId = 12345;
    const actions = {
      dispatchGetSingleItune: { type: songDetailContainerTypes.REQUEST_GET_SINGLE_ITUNE, trackId }
    };

    const props = mapDispatchToProps(submitSpy);
    props.dispatchGetSingleItune(trackId);
    await timeout(500);
    expect(submitSpy).toBeCalledWith(actions.dispatchGetSingleItune);
  });
});
