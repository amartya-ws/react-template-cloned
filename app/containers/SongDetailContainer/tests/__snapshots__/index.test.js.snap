// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SongDetailContainer /> container tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="SongDetailContainer__DetailsCard-sc-1v7fg8-0 liHVEB"
    >
      <img
        data-testid="song-image"
      />
      <div>
        <p
          class="T__StyledText-gjlic1-0 kYTaGb"
          data-testid="artist-name"
        >
          Artist name : 
        </p>
        <p
          class="T__StyledText-gjlic1-0 kYTaGb"
          data-testid="track-name"
        >
          Track name : 
        </p>
      </div>
    </div>
  </div>
</body>
`;
